### PR TITLE
feat: Search service adding suport for AWS Elasticsearch 

### PR DESCRIPTION
**Pull request by jsnowacki**
### Summary of Changes

PR adds support for AWS Elasticsearch, which requires different authentication for Elasticsearch client and some additional required libraries for AWS. The feature is discussed in #1135.

### Tests

Similar tests added as for `ElasticsearchProxy`, mosty inherited.

### Documentation

None so far.

### CheckList
Make sure you have checked **all** steps below to ensure a timely review.
- [x] PR title addresses the issue accurately and concisely. Example: "Updates the version of Flask to v1.0.2"
    - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
- [x] PR includes a summary of changes.
- [x] PR adds unit tests, updates existing unit tests, __OR__ documents why no test additions or modifications are needed.
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - All the public functions and the classes in the PR contain docstrings that explain what it does

### CODE REVIEW

**Comment by feng-tao** [comment]
there are some ongoing discussion in https://github.com/amundsen-io/amundsen/issues/1135

File: search/requirements.txt [hunk 1]
```diff
@@ -22,6 +23,7 @@ pytest==3.5.1
 pytest-cov==2.5.1
 pytest-mock==1.1
 pytz==2021.1
+requests-aws4auth==1.0.1
```
Comment (sewardgw) [inline search/requirements.txt:26]: Is it worth adding the aws reqs under some specific "aws" key within `extras_require`? I realize this convention hasn't been followed well for the search service but it seems like the correct approach.

File: search/search_service/proxy/elasticsearch.py [hunk 2]
```diff
@@ -667,3 +670,103 @@ def _get_mapping(alias: str) -> str:
         index_actions = {'actions': [{'add': {'index': index_key, 'alias': alias}}]}
         self.elasticsearch.indices.update_aliases(index_actions)
         return index_key
+
+
+def create_new_es_client_on_error(func):
```
Comment (sewardgw) [inline search/search_service/proxy/elasticsearch.py:675]: nit: may be worth adding `aws` in this name

File: search/requirements.txt (hunk 1 above)
Comment (mgorsk1) [reply search/requirements.txt:26]: there is a follow up discussion under https://github.com/amundsen-io/amundsen/issues/1135 @sewardgw

**Comment by jsnowacki** [comment]
There is some issue it seems with AWS credentials, but I setup env variables and the tests pass locally on my machine fine even if I remove credentials, whereas, they fail on the test runner, which I cannot reproduce it. Any ideas how to check this out or fix it?

**Comment by jsnowacki** [comment]
> There is some issue it seems with AWS credentials, but I setup env variables and the tests pass locally on my machine fine even if I remove credentials, whereas, they fail on the test runner, which I cannot reproduce it. Any ideas how to check this out or fix it?

OK I've found it, it was a spelling error of `requests-aws4auth` version.

File: search/search_service/config.py [hunk 3]
```diff
@@ -58,3 +58,30 @@ class LocalConfig(Config):
         'title': 'Search Service',
         'uiversion': 3
     }
+
+
+class AwsSearchConfig(LocalConfig):
+    import boto3
```
Comment (mgorsk1) [inline search/search_service/config.py:64]: in terms of docs I think we could just add a docstring to this class. correct me if I am wrong but `AwsSearchConfig` is prepared for one of many possible deployment strategies of amundsen on AWS - the one where you use AWS Elasticsearch service and authenticate containers to AWS ES service using IAM roles. The docstring should be explicit about the fact that it's only for this use-case.

Comment (jsnowacki) [reply search/search_service/config.py:99]: @mgorsk1 I've added docstring; please check if it's fine.

File: search/setup.cfg [hunk 4]
```diff
@@ -11,6 +11,10 @@ max-line-length = 120
 # modify --cov-fail-under parameter after adding unit tests
 [tool:pytest]
 addopts = --cov=search_service --cov-fail-under=0 --cov-report=term-missing:skip-covered --cov-report=xml --cov-report=html -vvv
+env =
+    AWS_REGION=eu-west-1
+    AWS_ACCESS_KEY_ID=foo
```
Comment (mgorsk1) [inline search/setup.cfg:16]: why do we need this? manual doesn't mention those vars so should they really be required for tests?

**Review by mgorsk1 — APPROVED** [review]
looks good, added one comment but if it doesn't make sense just disregard it

File: search/search_service/config.py (hunk 3 above)
Comment (mgorsk1) [reply search/search_service/config.py:99]: perfect, thanks!

File: search/setup.cfg (hunk 4 above)
Comment (jsnowacki) [reply search/setup.cfg:16]: `Session().get_credentials()` looks for them in [a couple of places](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/credentials.html#configuring-credentials) and env variables are the easiest to mock for it not to complain.

**Comment by mgorsk1** [comment]
@jsnowacki the change broke search proxy if aws proxy is not used, would you mind taking a look:

```
~/PycharmProjects/amundsen-god(feat/unify_deps*) » docker logs -f amundsensearch                                                                                            mariusz@mariusz-XPS-13-7390-2-in-1
Traceback (most recent call last):
  File "search_service/search_wsgi.py", line 6, in <module>
    from search_service import create_app
  File "/app/search_service/__init__.py", line 17, in <module>
    from search_service.api.dashboard import SearchDashboardAPI, SearchDashboardFilterAPI
  File "/app/search_service/api/dashboard.py", line 11, in <module>
    from search_service.api.base import BaseFilterAPI
  File "/app/search_service/api/base.py", line 12, in <module>
    from search_service.proxy import get_proxy_client
  File "/app/search_service/proxy/__init__.py", line 9, in <module>
    from search_service import config
  File "/app/search_service/config.py", line 63, in <module>
    class AwsSearchConfig(LocalConfig):
  File "/app/search_service/config.py", line 112, in AwsSearchConfig
    aws_auth = AWS4Auth(region=region, service=service, refreshable_credentials=credentials)
  File "/usr/local/lib/python3.7/site-packages/requests_aws4auth/aws4auth.py", line 284, in __init__
    raise TypeError(msg)
TypeError: AWS4Auth() takes 2, 4 or 5 arguments, 0 given

```

**Comment by jsnowacki** [comment]
I've commented in #1167

**Pull request merged**